grpc-server/src/main/java/com/netflix/conductor/grpc/server/service/GRPCHelper.java

```diff
@@ -0,0 +1,116 @@
+     *
+     * Note that this is trickier than it ought to be because the GRPC APIs have
+     * not been upgraded yet. Here's a quick breakdown of how this works in practice:
+     *
+     * Reporting a "status" result back to a client with GRPC is pretty straightforward.
+     * GRPC implementations simply serialize the status into several HTTP/2 trailer headers that
+     * are sent back to the client before shutting down the HTTP/2 stream.
+     *
+     * - 'grpc-status', which is a string representation of a {@link com.google.rpc.Code}
+     * - 'grpc-message', which is the description of the returned status
+     * - 'grpc-status-details-bin' (optional), which is an arbitrary payload with a serialized
+     *  ProtoBuf object, containing an accurate description of the error in case the status is not
+     *  successful.
+     *
+     *  By convention, Google provides a default set of ProtoBuf messages for the most common
+     *  error cases. Here, we'll be using {@link DebugInfo}, as we're reporting an internal
+     *  Java exception which we couldn't properly handle.
+     *
+     *  Now, how do we go about sending all those headers _and_ the {@link DebugInfo} payload
+     *  using the Java GRPC API?
+     *
+     *  The only way we can return an error with the Java API is by passing an instance of
+     *  {@link io.grpc.StatusException} or {@link io.grpc.StatusRuntimeException} to
+     *  {@link StreamObserver#onError(Throwable)}. The easiest way to create either of these
+     *  exceptions is by using the {@link Status} class and one of its predefined code
+     *  identifiers (in this case, {@link Status#INTERNAL} because we're reporting an internal
+     *  exception). The {@link Status} class has setters to set its most relevant attributes,
+     *  namely those that will be automatically serialized into the 'grpc-status' and 'grpc-message'
+     *  trailers in the response. There is, however, no setter to pass an arbitrary ProtoBuf message
+     *  to be serialized into a `grpc-status-details-bin` trailer. This feature exists in the other
+     *  language implementations but it hasn't been brought to Java yet.
+     *
+     *  Fortunately, {@link Status#asException(Metadata)} exists, allowing us to pass any amount
+     *  of arbitrary trailers before we close the response. So we're using this API to manually
+     *  craft the 'grpc-status-detail-bin' trailer, in the same way that the GRPC server implementations
+     *  for Go and C++ craft and serialize the header. This will allow us to access the metadata
+     *  cleanly from Go and C++ clients by using the 'details' method which _has_ been implemented
+     *  in those two clients.
+     *
+     * @param t The exception to convert
+     * @return an instance of {@link StatusException} which will properly serialize all its
+     * headers into the response.
+     */
+    private StatusException throwableToStatusException(Throwable t) {
```

Comment:
exception is better done via a gRPC interceptor

Reply:
I haven't found any good examples on this in the wild, so I decided to go with a helper method. Let me do some more research.

Reply:
To elaborate more on this: many (all?) of the methods in the DAOs we're using for the implementation are marked with `throws Exception`. This means that inside the `XXXServiceImplementation` for each method, we need an explicit `try {} catch` block, as it is not possible to add `throws Exception` to the actual RPC interface, as that would change the method's signature. Since we're handling the exception explicitly, I don't think it's possible to use an interceptor at all: we would need to re-throw the exception again, which would means changing the method's type signature.

Please advice.

Reply:
I need we need to fix the issue with "throws Exception", let's stick with what you have now and we will get back to this.